File: mobile/finance_track/lib/domain/entities/transaction.dart
import 'package:equatable/equatable.dart';

enum TransactionType { income, expense }

class Transaction extends Equatable {
  final int? id;
  final int walletId;
  final int categoryId;
  final TransactionType type;
  final double amount;
  final String description;
  final DateTime date;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Transaction({
    this.id,
    required this.walletId,
    required this.categoryId,
    required this.type,
    required this.amount,
    required this.description,
    required this.date,
    required this.createdAt,
    required this.updatedAt,
  });

  @override
  List<Object?> get props => [
        id,
        walletId,
        categoryId,
        type,
        amount,
        description,
        date,
        createdAt,
        updatedAt,
      ];

  Transaction copyWith({
    int? id,
    int? walletId,
    int? categoryId,
    TransactionType? type,
    double? amount,
    String? description,
    DateTime? date,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Transaction(
      id: id ?? this.id,
      walletId: walletId ?? this.walletId,
      categoryId: categoryId ?? this.categoryId,
      type: type ?? this.type,
      amount: amount ?? this.amount,
      description: description ?? this.description,
      date: date ?? this.date,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

File: apps/backend/api/v1/finance/serializers/transaction.py
from rest_framework import serializers
from finance.models import Transaction, Tag, TransactionTag  # tambahkan TransactionTag di sini
from .tag import TransactionTagSerializer

class TransactionSerializer(serializers.ModelSerializer):
    """
    Serializer untuk model Transaction (Transaksi Keuangan).
    
    Transaksi merepresentasikan pergerakan uang, bisa berupa pemasukan (income),
    pengeluaran (expense), atau transfer antar wallet.
    
    Attributes:
        wallet (int): ID wallet terkait
        wallet_name (str): Nama wallet (read-only)
        category (int): ID kategori transaksi
        category_name (str): Nama kategori (read-only)
        amount (decimal): Jumlah transaksi
        type (str): Tipe transaksi ('income', 'expense', 'transfer')
        description (str): Deskripsi transaksi (opsional)
        transaction_date (date): Tanggal transaksi
        created_at (datetime): Waktu pembuatan (read-only)
        updated_at (datetime): Waktu update terakhir (read-only)
        tags (list): Daftar tag terkait (read-only)
        tag_ids (list): Daftar ID tag untuk ditambahkan (write-only)
    """
    category_name = serializers.CharField(source='category.name', read_only=True)
    wallet_name = serializers.CharField(source='wallet.name', read_only=True)
    tags = TransactionTagSerializer(source='transaction_tags', many=True, read_only=True)
    tag_ids = serializers.ListField(
        child=serializers.IntegerField(),
        write_only=True,
        required=False
    )
    
    class Meta:
        model = Transaction
        fields = ['id', 'wallet', 'wallet_name', 'category', 'category_name', 
                  'amount', 'type', 'description', 'transaction_date', 
                  'created_at', 'updated_at', 'tags', 'tag_ids']
        read_only_fields = ['created_at', 'updated_at', 'user']

    def create(self, validated_data):
        tag_ids = validated_data.pop('tag_ids', [])
        user = self.context['request'].user
        validated_data['user'] = user
        
        transaction = super().create(validated_data)
        
        # Add tags
        for tag_id in tag_ids:
            try:
                tag = Tag.objects.get(id=tag_id, user=user)
                TransactionTag.objects.create(transaction=transaction, tag=tag)
            except Tag.DoesNotExist:
                pass
                
        return transaction
    
    def update(self, instance, validated_data):
        tag_ids = validated_data.pop('tag_ids', None)
        transaction = super().update(instance, validated_data)
        
        # Update tags only if provided
        if tag_ids is not None:
            user = self.context['request'].user
            
            # Clear existing tags
            TransactionTag.objects.filter(transaction=transaction).delete()
            
            # Add new tags
            for tag_id in tag_ids:
                try:
                    tag = Tag.objects.get(id=tag_id, user=user)
                    TransactionTag.objects.create(transaction=transaction, tag=tag)
                except Tag.DoesNotExist:
                    pass
                    
        return transaction


class TransactionListSerializer(serializers.ModelSerializer):
    """
    Lightweight serializer untuk model Transaction dalam format list.
    
    Digunakan untuk menampilkan daftar transaksi dengan informasi minimal
    yang diperlukan untuk list view.
    """
    
    category_name = serializers.CharField(source='category.name', read_only=True)
    wallet_name = serializers.CharField(source='wallet.name', read_only=True)
    
    class Meta:
        model = Transaction
        fields = ['id', 'wallet_name', 'category_name', 'amount', 
                  'type', 'transaction_date', 'description']